server/src/main/java/org/opensearch/snapshots/SnapshotsService.java

```diff
@@ -403,6 +448,175 @@ public TimeValue timeout() {
+                    shardGenerations,
+                    repositoryData.getGenId(),
+                    metadataForSnapshot(currentState.metadata(), request.includeGlobalState(), false, dataStreams, indexIds),
+                    snapshotInfo,
+                    version,
+                    state -> stateWithoutSnapshot(state, snapshot),
+                    new ActionListener<RepositoryData>() {
+                        @Override
+                        public void onResponse(RepositoryData repositoryData) {
+                            if (!clusterService.state().nodes().isLocalNodeElectedClusterManager()) {
+                                failSnapshotCompletionListeners(
+                                    snapshot,
+                                    new SnapshotException(snapshot, "Aborting Snapshot, no longer cluster manager")
+                                );
+                                listener.onFailure(
+                                    new SnapshotException(repositoryName, snapshotName, "Aborting Snapshot, no longer cluster manager")
+                                );
+                                return;
+                            }
+                            updateSnapshotPinnedTimestamp(repositoryData, snapshot, pinnedTimestamp, pinnedTimestampListener);
+                        }
+
+                        @Override
+                        public void onFailure(Exception e) {
+                            logger.error("Failed to upload files to snapshot repo {} for snapshot {} ", repositoryName, snapshotName);
+                            listener.onFailure(e);
+                        }
+                    }
+                );
+
+            }, listener::onFailure);
+        } catch (Exception e) {
+            assert false : new AssertionError(e);
+            logger.error("Snapshot {} creation failed with exception {}", snapshot.getSnapshotId().getName(), e);
+            listener.onFailure(e);
+        }
+    }
+
+    private void updateSnapshotPinnedTimestamp(
+        RepositoryData repositoryData,
+        Snapshot snapshot,
+        long timestampToPin,
+        ActionListener<RepositoryData> listener
+    ) {
+        listener.onResponse(repositoryData);
+        // remoteStorePinnedTimestampService.pinTimestamp(
+        // timestampToPin,
+        // snapshot.getRepository() + "__" + snapshot.getSnapshotId(),
+        // new ActionListener<Void>() {
+        // @Override
+        // public void onResponse(Void unused) {
+        // logger.debug("Timestamp pinned successfully for snapshot {}", snapshot.getSnapshotId().getName());
+        // listener.onResponse(repositoryData);
+        // }
+        //
+        // @Override
+        // public void onFailure(Exception e) {
+        // logger.error("Failed to pin timestamp for snapshot {} with exception {}", snapshot.getSnapshotId().getName(), e);
+        // listener.onFailure(e);
+        //
+        // }
+        // }
+        // );
```

Comment:
Remove?

Reply:
This is to pin the timestamp using RemoteStorePinnedTimestampService. Since the changes are merged now, will uncomment